terraform/monitoring/03_service_slo.tf

```diff
@@ -0,0 +1,100 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Here we create a custom service that we attach our SLOs to
+# Using a custom service here allows us to add on additional SLOs and 
+# alerting policies on things such as custom metrics
+resource "google_monitoring_custom_service" "frontend" {
+  service_id = "frontend-srv"
+  display_name = "Frontend Service"
+}
+
+# Here we define an SLO for availablity for the frontend service 
+# The SLO is defined as following: 
+#   80% of all non-4XX requests within the past 30 day windowed period
+#   return with 200 OK status
+resource "google_monitoring_slo" "frontend_availability_slo" {
+  service = google_monitoring_custom_service.frontend.service_id
+  slo_id = "availability-slo"
+  display_name = "Availability SLO with request base SLI (good total ratio)"
+
+  # The goal sets our 80% objective with a 30 day rolling window period
+  goal = 0.8
+  rolling_period_days = 30
+
+  request_based_sli {
+    good_total_ratio {
+
+      # The "good" service is the number of 200 OK responses
+      good_service_filter = join(" AND ", [
+        "metric.type=\"istio.io/service/server/request_count\"",
+        "resource.type=\"k8s_container\"",
+        "resource.label.\"cluster_name\"=\"stackdriver-sandbox\"",
+        "metadata.user_labels.\"app\"=\"frontend\"",
+        "metric.label.\"source_workload_namespace\"=\"default\"",
+        "metric.label.\"response_code\"=\"200\""
+      ])
+
+      # The total is the number of non-4XX responses
+      # We elimiate 4XX responses since they do not accurately represent server-side 
+      # failures and have the possibility of skewing our SLO measurements
+      total_service_filter = join(" AND ", [
+        "metric.type=\"istio.io/service/server/request_count\"",
+        "resource.type=\"k8s_container\"",
+        "resource.label.\"cluster_name\"=\"stackdriver-sandbox\"",
+        "metadata.user_labels.\"app\"=\"frontend\"",
+        "metric.label.\"source_workload_namespace\"=\"default\"",
+        "metric.label.\"response_code\"!=\"400\"",
+        "metric.label.\"response_code\"!=\"404\"",
+        "metric.label.\"response_code\"!=\"426\""
```

Comment:
Are these just the common 4XX errors that some documentation said to exclude?  There is no way to any kind of regex or prefix based matching so all 4XX errors can be excluded?

Reply:
Unfortunately the regex/prefix matching only works on strings and here it's stored as an int or number. I moved it to use a < || >= filter on the numbers that looks a lot cleaner.